src/model/reference.ml

```diff
@@ -170,43 +177,95 @@ let tokenize location s : token list =
+       no longer meaningful. *)
+    match s.[index] with
+    | exception Invalid_argument _ -> path_ended started_at index :: tokens
+    | '/' -> scan_path index (index - 1) (path_ended started_at index :: tokens)
+    | _ -> scan_path started_at (index - 1) tokens
+  and path_ended started_at index =
+    let offset = index + 1 in
+    let length = started_at - offset in
+    let identifier = String.sub s offset length in
+    let location = Location_.in_string s ~offset ~length location in
+    { kind = `End_in_slash; identifier; location }
   in
 
   scan_identifier (String.length s) 0 (String.length s - 1) [] |> List.rev
 
-let expected allowed location =
-  let unqualified = "or an unqualified reference" in
+let expected ?(expect_paths = false) allowed location =
+  let unqualified = [ "an unqualified reference" ] in
+  let unqualified =
+    if expect_paths then "a path" :: unqualified else unqualified
+  in
   let allowed =
-    match allowed with
-    | [ one ] -> Printf.sprintf "'%s-' %s" one unqualified
-    | _ ->
-        String.concat ", "
-          (List.map (Printf.sprintf "'%s-'") allowed @ [ unqualified ])
+    String.concat ", " (List.map (Printf.sprintf "'%s-'") allowed @ unqualified)
   in
   expected_err allowed location
 
+(* Parse references that do not contain a [/]. Raises errors and warnings. *)
 let parse whole_reference_location s :
     Paths.Reference.t Error.with_errors_and_warnings =
   let open Paths.Reference in
   let open Names in
+  let rec path components next_token tokens : Hierarchy.t =
+    match (next_token.kind, tokens) with
+    | `End_in_slash, [] when next_token.identifier = "" ->
+        (* {!/identifier} *)
+        (`TAbsolutePath, components)
+    | `End_in_slash, [] when next_token.identifier = "." ->
+        (* {!./identifier} *)
+        (`TRelativePath, components)
+    | `End_in_slash, [ { kind = `End_in_slash; identifier = ""; _ } ]
+      when next_token.identifier = "" ->
+        (* {!//identifier} *)
+        (`TCurrentPackage, components)
+    | `End_in_slash, [] ->
+        (* {!identifier'/identifier} *)
+        (`TRelativePath, next_token.identifier :: components)
+    | `End_in_slash, next_token' :: tokens' ->
+        (* {!path/identifier} *)
+        path (next_token.identifier :: components) next_token' tokens'
+    | (`None | `Prefixed _), _ ->
+        (* This is not really expected *)
+        expected ~expect_paths:true [] next_token.location
+        |> Error.raise_exception
+  in
+
+  let ends_in_slash next_token =
+    match next_token.kind with `End_in_slash -> true | _ -> false
+  in
+
   let rec signature { kind; identifier; location } tokens : Signature.t =
     let kind = match_reference_kind location kind in
     match tokens with
     | [] -> (
         match kind with
         | (`TUnknown | `TModule | `TModuleType) as kind ->
             `Root (identifier, kind)
+        | `TPathComponent -> `Module_path (`TRelativePath, [ identifier ])
```

Comment:
Why is that a relative path, even if `identifier = ""` ?
I think module paths must always be of this form:

`{!/libname/module-M.type-t}`

Reply:
I do think that indeed, `{!/M.type-t}` would yield a `` `Type( ...( `Module_path (`TRelativePath, [""])...) `` !

Reply:
Sorry this second sentence is answering an other comment. `identifier` cannot be `""` here and there's now an `assert false` in this case.

`{!/M.type-t}` parses to `` "`Type":[{"`Module_path":["`TAbsolutePath",["M"]]},"t"] ``.

Reply:
It took me a while but yes, I'm convinced this is indeed an `assert false`! Thanks for the explanation.

Tricky parser, good luck to the maintainers.